gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -253,13 +267,35 @@ public void run() {
                   return;
                 }
 
+                InsertObjectRequest.Builder requestBuilder = null;
+                // Create new request builder if the writeOffset grows over inserts. Otherwise
+                // something went wrong and resume write from a cached request builder.
+                if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+                  requestBuilder = buildRequestFromCachedDataChunk(dataChunkMap);
+                } else {
+                  requestBuilder = buildRequestFromPipeData();
+                }
+
+                if (requestBuilder == null) {
+                  return;
+                }
+
+                requestObserver.onNext(requestBuilder.build());
+                requestObserver.onCompleted();
+              }
+
+              // Handles the case where we're not resuming a write, and instead we need to read data
+              // from the input pipe.
+              private InsertObjectRequest.Builder buildRequestFromPipeData() {
                 try {
                   chunkData = readRequestData();
                 } catch (IOException e) {
-                  error =
-                      new IOException(
-                          String.format("Failed to read chunk for '%s'", resourceId), e);
-                  return;
+                  nonTransientError =
+                      new RuntimeException(
+                          String.format(
+                              "InsertChunkResponseObserver.beforeStart for uploadId %s caught %s",
+                              uploadId, e));
```

Comment:
Pass `e` as a parameter of the `RuntimeException` not `format` method, so its stack trace will be preserved.

Reply:
Done.